don't update parent event changed|failed in bulk (it's expensive for large tables)

ryanpetrello opened the pull request.

ryanpetrello committed:
don't update parent event changed|failed in bulk (it's expensive)

ryanpetrello commented on awx/main/models/events.py:
```diff
@@ -352,9 +352,19 @@ def save(self, *args, **kwargs):
+                if self.failed is True:
+                    kwargs['failed'] = True
+                if kwargs:
+                    JobEvent.objects.filter(job=self.job, uuid=self.parent_uuid).update(**kwargs)
```
```sql
\d main_jobevent

...
Indexes:
    ...
    "main_jobevent_job_id_uuid_3df694c5_idx" btree (job_id, uuid)
```

ryanpetrello commented on awx/main/models/events.py:
```diff
@@ -352,9 +352,19 @@ def save(self, *args, **kwargs):
+                if self.failed is True:
+                    kwargs['failed'] = True
+                if kwargs:
+                    JobEvent.objects.filter(job_id=self.job_id, uuid=self.parent_uuid).update(**kwargs)
```
```sql
\d main_jobevent

...
Indexes:
    ...
    "main_jobevent_job_id_uuid_3df694c5_idx" btree (job_id, uuid)
```

ryanpetrello replied:
Sample query observed running this through a debugger:

```sql
UPDATE "main_jobevent"
    SET "failed" = true
    WHERE ("main_jobevent"."job_id" = 7
    AND "main_jobevent"."uuid" = '0242ac12-0005-4e0c-d27c-00000000000d');
```

ryanpetrello replied:
Sample index scan on a production install:

```sql
awx=> EXPLAIN UPDATE "main_jobevent" SET "failed" = true WHERE ("main_jobevent"."job_id" = 7 AND "main_jobevent"."uuid" = '0242ac12-0005-4e0c-d27c-00000000000d');
                                                      QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------
 Update on main_jobevent  (cost=0.41..8.43 rows=1 width=964)
   ->  Index Scan using main_jobevent_job_id_1a69b01a6e969026_idx on main_jobevent  (cost=0.41..8.43 rows=1 width=964)
         Index Cond: ((job_id = 7) AND ((uuid)::text = '0242ac12-0005-4e0c-d27c-00000000000d'::text))
(3 rows)

awx=> SELECT COUNT(id) FROM main_jobevent;
 count
-------
 34765
(1 row)
```

matburt commented on awx/main/models/events.py:
```diff
@@ -352,9 +352,19 @@ def save(self, *args, **kwargs):
                 self._update_hosts()
-            if self.event == 'playbook_on_stats':
-                self._update_parents_failed_and_changed()
+            if any([
```
I think we can shortcut this entirely by just checking `self.changed` and `self.failed` for `True` and then falling into the JobEvent filter.

We'll save potentially `dozens` of microseconds!

matburt replied:
Even if this means affecting a parent above the task level I don't think that's such a big deal.

matburt commented:
One other thing will be checking to see if we bubble changed/failed up to a higher level beyond the task level in the current implementation.

ryanpetrello commented:
> One other thing will be checking to see if we bubble changed/failed up to a higher level beyond the task level in the current implementation.

I wouldn't _expect_ it to, given that we're currently filtering on these events on `.startwith('runner_on')`:

https://github.com/ansible/awx/blob/fc94b3a943d20c82216898e58c979027d80343a3/awx/main/models/events.py#L404

It seems to me like removing the `self.event` string check would result in a behavior change.

ryanpetrello commented:
Old query is _definitely_ slow super slow for large table sizes:

```sql
awx-dev=> INSERT INTO main_jobevent
    (created, modified, uuid, parent_uuid, failed, changed, host_name, play, role, task, counter, event, job_id, event_data, start_line, end_line, stdout, verbosity, playbook)
    SELECT now(), now(), x, 'parent_uuid', false, false, '', '', '', '', 1, 'playbook_on_start', 1, '{"pid": 3, "playbook_uuid": "461fdef8-ed03-402a-9f90-d283961ec8e3", "playbook": "hello_world.yml"}', 0, 1, 'foo', 0, 'hello_world.yml'
    FROM generate_series(1, 100000000) AS x;

awx-dev=> SELECT COUNT(id) FROM main_jobevent;
-[ RECORD 1 ]---
count | 11000043
```

`<run a simple Job Template that emits events>`

```sql
\timing
UPDATE "main_jobevent"
    SET "failed" = true
    WHERE "main_jobevent"."uuid" IN (
        SELECT U0."parent_uuid" AS Col1
        FROM "main_jobevent" U0
        WHERE (U0."job_id" = 16 AND U0."event"::text LIKE 'runner\_on%' AND U0."failed" = true)
    );
UPDATE 1
Time: 5060.824 ms
```

ryanpetrello commented:
If you pump the event count up to 25-30M, it can get pretty bad:

```sql
UPDATE "main_jobevent"
    SET "failed" = true
    WHERE "main_jobevent"."uuid" IN (
        SELECT U0."parent_uuid" AS Col1
        FROM "main_jobevent" U0
        WHERE (U0."job_id" = 16 AND U0."event"::text LIKE 'runner\_on%' AND U0."failed" = true)
    );
UPDATE 1
Time: 31314.081 ms
```
